docs: add initial docs.md, update FAQ.md (#236)
vlang/vab: docs/docs.md

Please contribute to this document if you think something is missing or could be better.

# Contents

- [Introduction](#introduction)
- [The V to Android App Development Cycle](#the-v-to-android-app-development-cycle)
- [Using `vab` from the command line](#using-vab-from-the-command-line)
- [Using `vab` programmatically](#using-vab-programmatically)
- [Examples](#examples)
- [Compile V code to Android Compatible C code](#compile-v-code-to-android-compatible-c-code)
- [Compile C code to Android shared library (.so)](#compile-c-code-to-android-shared-library-so)
- [Find and Invoke NDK Compiler Manually](#find-and-invoke-ndk-compiler-manually)

# Introduction

Android is a, primarily, Java based Operating System - and to be able to develop apps for
*the majority* of devices on the market (2022), the primary option is to target Google-based
Android devices.

To be able to develop for Google-based devices, Google provide the *Android SDK*, and
for native developement (C/C++), they provide the *Android NDK*. Both are currently required
to be able to develop Android apps in V.

> While the Android OS itself is open source - the SDK and NDK distributed by Google is ***not***.

Reviewer: ```suggestion
While the Android OS itself is open source - the SDK and NDK distributed by Google are ***not***.
```

> This fact makes everything a bit harder than it ought to be.

Reviewer: That is an understatement, if there ever was one ...

So - for us to target the majority, we need the Android SDK and NDK that are
both proprietary licensed in such a way that they prevent developers from
modifying and re-distributing them - thus forcing developers to use Google's
official (rather large) distributions of these developement kits.

For historical reasons - the job of locating and using all the tools in the
development kits is actually harder than it might appear at first sight
since Google tend to move things around, deprecate tools on short, if any, notice
and, in general, always be in a kind of "beta" mode when they release the software
needed to target and build for the Google flavoured Android OS devices.

To compensate for this they've developed a ~900MiB tool called "Android Studio"
to aid developers in developing for their platform... Because of it's huge size,
slow compile speeds and the difficulties involved in integrating a satisfying user

> experience for, at least, V developers - V has chosen to develop a tool that

Reviewer: ```suggestion
experience for V developers - V has chosen to develop a tool that
```
but I am not sure actually - @JalonSolov what do you think?

completely eliminate the need to use Android Studio in V development.

With all this in mind `vab` was born to make life easier for V developers.
You can install it by following the instructions [in the README.md](https://github.com/vlang/vab#install).

> Because of the situation described above and the design of the Android developer eco-system,

Reviewer: ```suggestion
Because of the situation described above and the design of the Android developer ecosystem,
```

`vab` is basically just a wrapper around the Android SDK and NDK. `vab` is able to both
install, locate and orchestrate the tools found inside the distributed development kits.

> Welcome - and have a productive time using V and `vab`!

Reviewer: That may be put on top.

Author: Good idea

Author: Done in [ebdda19](https://github.com/vlang/vab/pull/236/commits/ebdda192e01e196ad7edd2c3b0ef9d2d19965e88) along side, mostly, formatting changes

1. Develop an app in V (using `$if android {...}` for Android specific code).
2. Compile the V code to Android compatible C code (`v -os android -o code.c ...`).
3. Compile Android compatible C code to a *shared* library (`.so`) with a NDK compiler.
4. Prepare all base files (e.g. `AndroidManifest.xml`, Java activity sources, etc.)
needed for the Android app to be able to start on the device OS.
5. Compile Java sources to load and bootstrap the *shared* library + any Java dependencies required.
6. Package everything up for distribution via an `.apk` or `.aab` archive
(dex sources, copy native libs etc.).
7. Sign the packaged `.apk`/`.aab` with a keytool.
8. (Optionally) deploy and/or run the app on an emulator or physical device.
9. (Optionally) debug the app (via `adb logcat ...`).

The developer is expected to take care of bullet point `1.` while `vab` will help cover the rest.

# Using `vab` from the command line

For now please use the [Usage](https://github.com/vlang/vab#usage) section of the README.md
and the [FAQ.md](https://github.com/vlang/vab/blob/master/docs/FAQ.md).

# Using `vab` programmatically

> First off; please keep in mind that using `vab` programmatically has the same

Reviewer: ```suggestion
Please keep in mind that using `vab` programmatically has the same
```

prerequisites as using the `vab` command line tool. Use `vab doctor` on the command line
to check if all dependencies are met and setup correctly.

You can also invoke the `doctor` programmatically:

```v ignore
import vab.cli

cli.doctor(...)
```

> On top, it's *recommended* that you have some experience with Android development already

Reviewer: ```suggestion
On top, it's *recommended* that you have some experience with Android development already,
```

as some terms and situations can be resolved more quickly with knowledge about the different
areas and implications of Android developement.

In some situations, calling `vab` from the command line might not fit
a specific project, setup or problem at hand. You usually want more control
and flexibility - fear not - this is why `vab` is also exposed as a module (`import vab.xxx`).

The most useful modules are the following:

```v ignore
import vab.cli          // For easy replication of the `vab` command line tool
import vab.android      // For invoking the major steps: compile, package and deploy
import vab.android.sdk  // For easy access to tools in the Android SDK
import vab.android.ndk  // For easy access to tools in the Android NDK
```

> For general programmatic usage see the, relatively, small [`vab.v`](https://github.com/vlang/vab/blob/master/vab.v) that makes up the command line tool,

Reviewer: ```suggestion
For general programmatic usage see the relatively small [`vab.v`](https://github.com/vlang/vab/blob/master/vab.v), that makes up the command line tool,
```

> it shows the *major* steps you need to take to get from V source code to a running Android app.

Reviewer: ```suggestion
It shows the *major* steps you need to take, to get from V source code to a running Android app.
```

Working with native code like C (and implicitly V) on Android can get a little
cumbersome when things doesn't fit into the usual boxes, especially because the smallest
things can often touch and affect *all* the different stages in developing an

> Android application. `vab` is designed with this in mind to be able handle each of

Reviewer: ```suggestion
Android application. `vab` is designed with this in mind, to be able handle each of
```

these stages as flexible and easy as possible.

By using `vab` as a module you get access to both the high-level operations
(compile, package, deploy) as well as fine grained control if you, for example,
just want to compile a C file with a specific NDK compiler or locate the path(s) to

> the SDK or NDK on the user's host computer.

Reviewer: ```suggestion
the SDK or NDK on your host computer.
```

[`android/deploy.v`](https://github.com/vlang/vab/blob/master/android/deploy.v).

If you're in doubt how some call is supposed to work - the source code of
these functions is a good place to discover *how* things, in general, are
used and in what *order* they should be invoked.

They basically contain a function for each *major* step of the process:
```v ignore
import vab.android

// Compile V source to C source, and C source to a shared library
compile_opt := android.CompileOptions{
    ...
}
android.compile(compile_opt) or { panic(err) }

// Prepare base files, compile Java sources, package everything up in a .apk/.aab
package_opt := android.PackageOptions{
    ...
}
android.package(package_opt) or { panic(err) }

// Deploy the package to an emulator or device
deploy_opt := android.DeployOptions{
    ...
}
android.deploy(deploy_opt) or { panic(err) }
```

# Examples

The following are some useful examples, please contribute to this section if you think something
is missing or could be better. All examples assumes that you have *a working* install of `vab`.

## Compile V code to Android Compatible C code

> In this section it's important to keep in mind that most (if not all?) C libs that control the

Reviewer: ```suggestion
It's important to keep in mind that most (if not all?) C libs that control the
```

app window and (OpenGL) acceleration context usually have framework-like behaviour on Android, which
means that they usually require full authority over the way they start up and open the
[NativeActivity](https://developer.android.com/ndk/reference/group/native-activity) which is the entry point for *native* graphical apps.

This is often the reason why a lot of fiddling is required to get C libs like `SDL2`, `sokol_app`
and `RayLib` mixed with V code; All of them want control over how the window/context is opened.

Luckily V allows for outputting code which is agnostic to who or what will open the window.

> From the command line this is basically just invoking `v` with the right set of flags for Android:

Reviewer: ```suggestion
From the command line, this is basically just invoking the `v` compiler with the right set of flags for Android:
```
